content/guides/typescript-for-solid.mdx

```diff
@@ -0,0 +1,714 @@
+This resolves to a getter-setter [tuple](https://www.typescriptlang.org/docs/handbook/2/objects.html#tuple-types), both of which are generically typed:
+
+```typescript
+import type { Signal, Accessor, Setter } from "solid-js";
+type Signal<T> = [get: Accessor<T>, set: Setter<T>];
+```
+
+In this case, the `count` signal getter has the type `Accessor<number | undefined>`. 
+In Solid, `Accessor<T>` is equivalent to `() => T`. 
+Here, it means `count` is a function that returns either a `number` or `undefined`.
+
+The reason `undefined` is included in the type is because the signal was not initialized. 
+Therefore, its initial state can be `undefined`.
+
+The signal setter, `setCount`, is of type `Setter<number | undefined>`. 
+This complex type roughly resembles `(value? : number | ((prev?: number) => number)) => number`, accommodating both a direct number or a function which takes a potential pervious value that results in a number.
+Since the signal's start is undefined, both the direct number and the function's number parameter are optional.
+
+When a signal is designed to store functions, its setter function won't directly accept new functions as values.
+This is because the setter can't determine whether the function should be executed to get the actual value. 
+In these situations, you should use the callback form of the setter: `setSignal(() => value)`.
+
+
+#### Default values
+
+By providing default values when you call `createSignal`, you can avoid the need for explicit type specification and eliminate the possibility of the `| undefined` type. 
+This leverages type inference to determine the type automatically:
+
+```typescript
+const [count, setCount] = createSignal(0);
+const [name, setName] = createSignal("");
+```
+
+In this example, TypeScript understands the types as `number` and `string`.
+This means that `count` and `name` directly receive the types `Accessor<number>` and `Accessor<string>`, respectively, without the `| undefined` tag.
+
+### Context
+
+Just as signals utilize `createSignal<T>`, context uses `createContext<T>`, which is parameterized by the type `T` of the context's value:
+
+```tsx
+type Data = { count: number; name: string };
+```
+
+When you invoke `useContext(dataContext)`, it returns the type contained within the context. 
+For example, if the context is of type `Context<Data | undefined>`, then using `useContext` with it will naturally yield a result of type `Data | undefined`. 
+The `| undefined` part signifies that the context might not be defined in the component's ancestor hierarchy.
+`dataContext` is understood as `Context<Data | undefined>`. 
+Invoking `useContext(dataContext)` mirrors this type, returning `Data | undefined`.
+The `| undefined` arises when context is not defined in the component's ancestors, which leads `useContext` to yield `undefined`.
```

Comment:
```suggestion
The `| undefined` arises since it cannot be determined if the context's value will be used in a context where it has been assigned. 
```